File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_typescript-vite/src/lib/handlers/quests.ts
import type { ServerResponse } from 'node:http';
import { sendError, sendJson } from '../http.js';
import { campaignExists, createQuest, getQuest, getQuestsByCampaign, questExists, updateQuest } from '../db.js';
import { isNonEmptyString, isQuestStatus, isStringArray } from '../validation.js';
import type { Quest, QuestStatus } from '../types.js';

function milestonesDone(quest: Quest): number {
  const valid = new Set(quest.milestones);
  const unique = new Set<string>();
  for (const m of quest.completed) {
    if (valid.has(m)) unique.add(m);
  }
  return unique.size;
}

export function handleCreateQuest(pathname: string, body: unknown, res: ServerResponse): boolean {
  const match = pathname.match(/^\/v1\/campaigns\/(.+)\/quests$/);
  if (!match) return false;
  const campaignId = match[1];
  if (!campaignExists(campaignId)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  const b = body as any;
  if (!b || !isNonEmptyString(b.id) || !isNonEmptyString(b.title) || !isQuestStatus(b.status) || !isStringArray(b.milestones)) {
    sendError(res, 400, 'invalid request');
    return true;
  }
  if (questExists(b.id)) {
    sendError(res, 409, 'quest already exists');
    return true;
  }
  const quest: Quest = {
    id: b.id,
    campaign_id: campaignId,
    title: b.title,
    status: b.status as QuestStatus,
    milestones: b.milestones,
    completed: [],
  };
  createQuest(quest);
  sendJson(res, 201, {
    id: quest.id,
    title: quest.title,
    status: quest.status,
    milestones_total: quest.milestones.length,
    milestones_done: 0,
  });
  return true;
}

export function handleUpdateQuestProgress(pathname: string, body: unknown, res: ServerResponse): boolean {
  const match = pathname.match(/^\/v1\/campaigns\/(.+)\/quests\/(.+)\/progress$/);
  if (!match) return false;
  const campaignId = match[1];
  const questId = match[2];
  if (!campaignExists(campaignId)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  const quest = getQuest(questId);
  if (!quest || quest.campaign_id !== campaignId) {
    sendError(res, 404, 'quest not found');
    return true;
  }
  const b = body as any;
  if (!b || !isStringArray(b.completed)) {
    sendError(res, 400, 'invalid request');
    return true;
  }
  const completed = b.completed as string[];
  const valid = new Set(quest.milestones);
  const merged = new Set(quest.completed);
  for (const m of completed) {
    if (valid.has(m)) merged.add(m);
  }
  quest.completed = Array.from(merged);
  if (quest.completed.length >= quest.milestones.length && quest.milestones.length > 0) {
    quest.status = 'completed';
  }
  updateQuest(quest);
  const done = milestonesDone(quest);
  sendJson(res, 200, {
    id: quest.id,
    status: quest.status,
    milestones_total: quest.milestones.length,
    milestones_done: done,
  });
  return true;
}

export function handleGetQuestSummary(pathname: string, res: ServerResponse): boolean {
  const match = pathname.match(/^\/v1\/campaigns\/(.+)\/quests\/summary$/);
  if (!match) return false;
  const campaignId = match[1];
  if (!campaignExists(campaignId)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  const counts: Record<QuestStatus, number> = { active: 0, completed: 0, blocked: 0 };
  for (const quest of getQuestsByCampaign(campaignId)) {
    counts[quest.status]++;
  }
  sendJson(res, 200, {
    campaign_id: campaignId,
    active: counts.active,
    completed: counts.completed,
    blocked: counts.blocked,
  });
  return true;
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T061539Z_claude_claude-sonnet-5_rust-stdlib/src/quests.rs
//! Campaign quest tracking: create quests with a milestone checklist, mark
//! milestones complete, and summarize quest counts by status per campaign.
//! Backing store is an in-memory map keyed by campaign id (each campaign
//! owns a `Vec<Quest>`), separate from [`crate::campaigns`]'s own store but
//! validated against it the same way [`crate::dm_tools`] validates
//! `campaign_id`.

use std::collections::HashMap;
use std::net::TcpStream;
use std::sync::{Mutex, OnceLock};

use crate::campaigns::campaigns;
use crate::http::{bad_request, respond};
use crate::json::{escape_json_string, parse_json};

const STATUSES: [&str; 3] = ["active", "completed", "blocked"];

pub(crate) struct Quest {
    id: String,
    title: String,
    status: String,
    milestones: Vec<String>,
    done: Vec<String>,
}

pub(crate) fn quests() -> &'static Mutex<HashMap<String, Vec<Quest>>> {
    static QUESTS: OnceLock<Mutex<HashMap<String, Vec<Quest>>>> = OnceLock::new();
    QUESTS.get_or_init(|| Mutex::new(HashMap::new()))
}

pub(crate) fn clear() {
    quests().lock().unwrap().clear();
}

pub(crate) fn quest_count(campaign_id: &str) -> i64 {
    let store = quests().lock().unwrap();
    store.get(campaign_id).map(|qs| qs.len() as i64).unwrap_or(0)
}

/// Returns `(active, completed, blocked)` quest counts for a campaign.
pub(crate) fn status_counts(campaign_id: &str) -> (i64, i64, i64) {
    let store = quests().lock().unwrap();
    let mut active = 0_i64;
    let mut completed = 0_i64;
    let mut blocked = 0_i64;
    if let Some(campaign_quests) = store.get(campaign_id) {
        for q in campaign_quests {
            match q.status.as_str() {
                "active" => active += 1,
                "completed" => completed += 1,
                "blocked" => blocked += 1,
                _ => {}
            }
        }
    }
    (active, completed, blocked)
}

/// Quests that are not yet completed (active or blocked).
pub(crate) fn open_quest_count(campaign_id: &str) -> i64 {
    let (active, _completed, blocked) = status_counts(campaign_id);
    active + blocked
}

fn create_json(q: &Quest) -> String {
    format!(
        r#"{{"id":"{}","title":"{}","status":"{}","milestones_total":{},"milestones_done":{}}}"#,
        escape_json_string(&q.id),
        escape_json_string(&q.title),
        escape_json_string(&q.status),
        q.milestones.len(),
        q.done.len()
    )
}

fn progress_json(q: &Quest) -> String {
    format!(
        r#"{{"id":"{}","status":"{}","milestones_total":{},"milestones_done":{}}}"#,
        escape_json_string(&q.id),
        escape_json_string(&q.status),
        q.milestones.len(),
        q.done.len()
    )
}

pub(crate) fn handle_create_quest(
    stream: &mut TcpStream,
    campaign_id: &str,
    body: &str,
) -> std::io::Result<()> {
    let json = match parse_json(body) {
        Some(j) => j,
        None => return bad_request(stream, "invalid json"),
    };

    let id = match json.get("id").and_then(|v| v.as_str()) {
        Some(s) if !s.is_empty() => s.to_string(),
        _ => return bad_request(stream, "invalid id"),
    };
    let title = match json.get("title").and_then(|v| v.as_str()) {
        Some(s) if !s.is_empty() => s.to_string(),
        _ => return bad_request(stream, "invalid title"),
    };
    let status = match json.get("status").and_then(|v| v.as_str()) {
        Some(s) if STATUSES.contains(&s) => s.to_string(),
        _ => return bad_request(stream, "invalid status"),
    };
    let milestones_json = match json.get("milestones").and_then(|v| v.as_array()) {
        Some(a) => a,
        None => return bad_request(stream, "invalid milestones"),
    };
    let mut milestones = Vec::new();
    for m in milestones_json {
        match m.as_str() {
            Some(s) if !s.is_empty() => milestones.push(s.to_string()),
            _ => return bad_request(stream, "invalid milestones"),
        }
    }

    {
        let campaigns_store = campaigns().lock().unwrap();
        if !campaigns_store.contains_key(campaign_id) {
            return respond(stream, 404, r#"{"error":"campaign not found"}"#);
        }
    }

    let mut store = quests().lock().unwrap();
    let campaign_quests = store.entry(campaign_id.to_string()).or_insert_with(Vec::new);
    if campaign_quests.iter().any(|q| q.id == id) {
        return respond(stream, 409, r#"{"error":"quest already exists"}"#);
    }

    let quest = Quest {
        id,
        title,
        status,
        milestones,
        done: Vec::new(),
    };
    let out = create_json(&quest);
    campaign_quests.push(quest);
    drop(store);

    respond(stream, 201, &out)
}

pub(crate) fn handle_quest_progress(
    stream: &mut TcpStream,
    campaign_id: &str,
    quest_id: &str,
    body: &str,
) -> std::io::Result<()> {
    let json = match parse_json(body) {
        Some(j) => j,
        None => return bad_request(stream, "invalid json"),
    };

    let completed_json = match json.get("completed").and_then(|v| v.as_array()) {
        Some(a) => a,
        None => return bad_request(stream, "invalid completed"),
    };
    let mut completed = Vec::new();
    for m in completed_json {
        match m.as_str() {
            Some(s) if !s.is_empty() => completed.push(s.to_string()),
            _ => return bad_request(stream, "invalid completed"),
        }
    }

    let mut store = quests().lock().unwrap();
    let campaign_quests = match store.get_mut(campaign_id) {
        Some(qs) => qs,
        None => return respond(stream, 404, r#"{"error":"campaign not found"}"#),
    };
    let quest = match campaign_quests.iter_mut().find(|q| q.id == quest_id) {
        Some(q) => q,
        None => return respond(stream, 404, r#"{"error":"quest not found"}"#),
    };

    for milestone in &completed {
        if !quest.milestones.contains(milestone) {
            return bad_request(stream, "unknown milestone");
        }
    }
    for milestone in completed {
        if !quest.done.contains(&milestone) {
            quest.done.push(milestone);
        }
    }
    if !quest.milestones.is_empty() && quest.done.len() == quest.milestones.len() {
        quest.status = "completed".to_string();
    }

    let out = progress_json(quest);
    drop(store);

    respond(stream, 200, &out)
}

pub(crate) fn handle_quest_summary(stream: &mut TcpStream, campaign_id: &str) -> std::io::Result<()> {
    let campaigns_store = campaigns().lock().unwrap();
    if !campaigns_store.contains_key(campaign_id) {
        return respond(stream, 404, r#"{"error":"campaign not found"}"#);
    }
    drop(campaigns_store);

    let (active, completed, blocked) = status_counts(campaign_id);

    let out = format!(
        r#"{{"campaign_id":"{}","active":{},"completed":{},"blocked":{}}}"#,
        escape_json_string(campaign_id),
        active,
        completed,
        blocked
    );
    respond(stream, 200, &out)
}
